### PR TITLE
Block Arnoldi

### LINKED ISSUE #2340: Implement block Arnoldi
Currently, [`arnoldi`](https://github.com/pymor/pymor/blob/225e12abfb5f0985187e041c832b9a4326e88eb3/src/pymor/algorithms/krylov.py#L10) works only with a single vector. Extending this to multiple vectors would be of interest.

### PR DESCRIPTION
**Pull request by sdrave**
Fixes #2340.

### CODE REVIEW

File: src/pymor/algorithms/krylov.py [hunk 1]
```diff
@@ -38,21 +40,38 @@ def arnoldi(A, E, b, r):
         Orthonormal basis for the Krylov subspace as a |VectorArray|.
     """
     assert A.source == A.range
+    if E is None:
+        E = IdentityOperator(A.source)
```
Comment (pmli) [inline src/pymor/algorithms/krylov.py:44]: It should be documented that `E` is optional.

**Review by pmli — CHANGES_REQUESTED** [review]
Looks good to me, besides the docs remark.

File: src/pymor/algorithms/krylov.py (hunk 1 above)
Comment (drittelhacker) [reply src/pymor/algorithms/krylov.py:44]: Should we also mention that, while `r * len(b)` is an upper bound to the produced basis size, due to deflation the actual basis might be smaller?

File: src/pymor/algorithms/krylov.py [hunk 2]
```diff
@@ -38,21 +40,38 @@ def arnoldi(A, E, b, r):
...
+
+    V = A.source.empty(reserve=r*len(b))
 
-    V = A.source.empty(reserve=r)
     v = E.apply_inverse(b)
-    v.scal(1 / v.norm()[0])
-    V.append(v)
+    V.append(v, remove_from_other=True)
+    gram_schmidt(V, atol=0, rtol=0, copy=False)
+    if len(V) < r:
```
Comment (drittelhacker) [inline src/pymor/algorithms/krylov.py:57]:
```suggestion
    if len(V) < len(b):
```

Comment (sdrave) [reply src/pymor/algorithms/krylov.py:57]: Good catch, thanks.

File: src/pymor/algorithms/krylov.py (hunk 1 above)
Comment (sdrave) [reply src/pymor/algorithms/krylov.py:46]: Did both. I also removed the 'real' everywhere from the docstring, as I don't see what it is supposed to mean exactly. The algorithm would work the same for complex data, returning an orthonormal basis for the respective $\mathbb{C}$-Krylov space.

**Review by pmli — APPROVED** [review]

**Review by drittelhacker — APPROVED** [review]

**Pull request merged**